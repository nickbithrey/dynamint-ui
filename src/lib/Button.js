import React from 'react';
import PropTypes from 'prop-types';
import { 
	PrimaryButton, 
	DefaultButton, 
	CommandBarButton, 
	IconButton 
} from 'office-ui-fabric-react/lib/Button';

const buttonTypes = {
	standard: DefaultButton,
	primary: PrimaryButton,
	command: CommandBarButton,
	icon: IconButton
};

const Button = ({text, onClick, type}) => {
	const Btn = buttonTypes[type];
	return <Btn text={text} onClick={onClick} />
};

Button.defaultProps = {
	type: 'standard'
};

Button.propTypes = {
	text: PropTypes.string.isRequired,
	onClick: PropTypes.func
}

export default Button;

export const newBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'standard'
	}
);

export const newPrimaryBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'primary'
	}
);