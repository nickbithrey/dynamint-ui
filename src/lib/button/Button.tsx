import * as React from 'react';
import { 
    PrimaryButton, 
    DefaultButton, 
    CommandBarButton, 
    IconButton,
    BaseButton
} from 'office-ui-fabric-react/lib/Button';
import { Link } from 'react-router-dom';
import * as Assert from 'util/Assert';

export interface IButton {
    content: string;
    onClick?: (e?: React.MouseEvent<BaseButton>) => void;
    type: 'primary' | 'default' | 'command' | 'icon';
    link?: string;
}

const buttonTypes = {
    primary: PrimaryButton,
    default: DefaultButton,
    command: CommandBarButton,
    icon: IconButton
};

export class Button extends React.Component<IButton> {
    
    constructor(props: IButton) {
        super(props);
    }
    
    render() {
        const {
            content,
            onClick,
            type,
            link
        } = this.props;
        const Btn = buttonTypes[type];
        Assert.assertDefined(Btn, 'no button registered for type ' + type + '. only types ' + Object.keys(buttonTypes) + ' are available');
        const result = <Btn 
                    text={content}
                    onClick={onClick}
                />
        if (link) {
            return <Link to={link}>{result}</Link>
        }
        return result;
    }
    
}