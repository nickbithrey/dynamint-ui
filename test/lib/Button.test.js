import React from 'react';
import { shallow } from 'enzyme';

import Button, { newBtn, newPrimaryBtn } from '~/lib/Button';

const buttonTagName = 'CustomizedDefaultButton';

describe('<Button /> rendering', () => {

	it('can render standard button', () => {
		const props = {
			text: 'name',
			onClick: jest.fn(() => [])
		};
		const wrapper = shallow( <Button {...props} /> );
		
		expect(wrapper).toMatchSnapshot();
	});
	
	it('calls onClick function on when clicking', () => {
		const props = {
			text: 'name',
			onClick: jest.fn(() => [])
		};
		const wrapper = shallow( <Button {...props} /> );
		
		expect(wrapper.find(buttonTagName)).toHaveLength(1);
		expect(wrapper.find(buttonTagName).prop())
		wrapper.find(buttonTagName).simulate('click');
		console.log(wrapper.props);
		expect(wrapper.prop('onClick')).toBeCalled();
		expect(wrapper.prop('onClick').mock.calls.length).toBe(1);
	});
		
});
	
describe('functions for creating buttons', () => {
	
	test('new standard button function', () => {
		const name = 'name';
		const click = jest.fn();
		const condition = jest.fn();
		
		const btn = newBtn(name, click, condition);
		
		expect(btn.text).toBe(name);
		expect(btn.onClick).toBe(click);
		expect(btn.condition).toBe(condition);
		expect(btn.type).toBe('standard');
	});
	
	test('new primary button function', () => {
		const name = 'name';
		const click = jest.fn();
		const condition = jest.fn();
		
		const btn = newPrimaryBtn(name, click, condition);
		
		expect(btn.text).toBe(name);
		expect(btn.onClick).toBe(click);
		expect(btn.condition).toBe(condition);
		expect(btn.type).toBe('primary');
	});

});