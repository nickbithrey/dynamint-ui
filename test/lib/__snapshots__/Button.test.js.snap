// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> rendering can render standard button 1`] = `
<CustomizedDefaultButton
  onClick={[MockFunction]}
  text="name"
/>
`;
